targets/csharp/source/Plugins/CloudScript/source/PlayFabFunctionContexts.cs

```diff
@@ -0,0 +1,245 @@
+
+        public TFunctionArgument FunctionArgument { get; set; }
+
+        protected FunctionTaskContext() { }
+
+        /// <summary>
+        /// Creates a new <c>FunctionTaskContext</c> out of the incoming request to an Azure Function.
+        /// </summary>
+        /// <param name="request">The request incoming to an Azure Function from the PlayFab server</param>
+        /// <returns>A new populated <c>FunctionTaskContext</c> instance</returns>
+        public static async Task<FunctionTaskContext<TFunctionArgument>> Create(HttpRequestMessage request)
+        {
+            using (var content = request.Content)
+            {
+                var body = await content.ReadAsStringAsync();
+                var contextInternal = PlayFabSimpleJson.DeserializeObject<FunctionTaskContextInternal>(body);
+                var settings = new PlayFabApiSettings
+                {
+                    TitleId = contextInternal.TitleAuthenticationContext.Id,
+                    DeveloperSecretKey = Environment.GetEnvironmentVariable("PLAYFAB_DEV_SECRET_KEY", EnvironmentVariableTarget.Process)
+                };
+
+                var authContext = new PlayFabAuthenticationContext
+                {
+                    EntityToken = contextInternal.TitleAuthenticationContext.EntityToken
+                };
+
+                return new FunctionTaskContext<TFunctionArgument>()
+                {
+                    ApiSettings = settings,
+                    AuthenticationContext = authContext,
+                    ScheduledTaskNameId = contextInternal.ScheduledTaskNameId,
+                    EventHistory = contextInternal.EventHistory,
+                    FunctionArgument = contextInternal.FunctionArgument
+                };
+            }
+        }
+
+        private class FunctionTaskContextInternal
+        {
+            public NameIdentifier ScheduledTaskNameId { get; set; }
+
+            public Stack<PlayStreamEventHistory> EventHistory { get; set; }
+
+            public TitleAuthenticationContext TitleAuthenticationContext { get; set; }
+
+            public TFunctionArgument FunctionArgument { get; set; }
+        }
+    }
+
+    public class NameIdentifier
+    {
+        public string Name { get; set; }
+
+        public string Id { get; set; }
+    }
+
+    public class PlayStreamEventEnvelope
+    {
+        public string EntityId { get; set; }
+        public string EntityType { get; set; }
+        public string EventName { get; set; }
+        public string EventNamespace { get; set; }
+        public string EventData { get; set; }
+        public string EventSettings { get; set; }
+    }
+
+    public class PlayStreamEventHistory
+    {
+        public string ParentTriggerId { get; set; }
+
+        public string ParentEventId { get; set; }
+
+        public bool TriggeredEvents { get; set; }
+    }
+
+    internal class TitleAuthenticationContext
+    {
+        public string Id { get; set; }
```

Comment:
What is this ID? TitleId?

Reply:
Yep. It was general consensus from the MainServer side to not call this TitleId and TitleX TitleY TitleZ because the class itself has Title in its name.

Reply:
I'm on the fence for this one.
Let's add a summary for it, I think that'll ease my hesitation.

Reply:
Sure I can try to do that but this is an internal class that's not being exposed to the public interface of the library so I believe adding docs for it is rather unnecessary?

Reply:
It's simply a structure for typing what goes through the wire from mainserver to azure function

Reply:
Yes and no... I mean...
Our team has to maintain this code, and when things are unclear, we can help out our future selves with a bit of clarity.
I will not remember this conversation in 2 months (or in 2 weeks even...)